### ISSUE TITLE
GridColDef - Add a metadata property or allow a type override to extend the type

### ISSUE THREAD

**Issue by tomdglenn91** [mui/mui-x#11393]
### Search keywords

datagrid column definition metadata

### Latest version

- [X] I have tested the latest version

### Summary 💡

When providing a col definition, I should be able to define additional fields to use in other behaviours across datagrid. This will let me perform table or application unique behaviour in functions that pass in information about the column, or let me use it via apiRef, such as valueGetter, onRowClick, getRowClassName, etc. Passing in arbitrary data to use later will be very useful, but should hopefully not be destructive to the GridColDef API.

### Examples 🌈

```
const col: GridColDef<any, any, any, { bar: false }> = {
    field: 'foo',
    valueGetter: ({ colDef }) => colDef.metadata.bar ? 'a' : 'b'
    metadata: {
       bar: false
    }
}
```

or via typescript

```
declare module '@mui/x-data-grid-premium' {
  interface GridColDefOverrides {
     bar: false
  }
}

const col: GridColDef<any, any, any, { bar: false }> = {
    field: 'foo',
    valueGetter: ({ colDef }) => colDef.bar ? 'a' : 'b'
    bar: false    
}
```

### Motivation 🔦

I'm trying to wire in datagrid to our API, and some filters require transforming. For example, I may need to transform datagrid saying "foo contains ABC" into "foo=%ABC%" or other transformation. For keeping everything on the column definition, i'd like to append this transformation function to the columns itself, where a utility function can pick them up. So for example

```
const col: GridColDef<any, any, any, { bar: false }> = {
    field: 'foo',
    valueGetter: ({ colDef }) => colDef.metadata.bar ? 'a' : 'b'
    metadata: {
       filterTransform: (operator, value) => ['foo', operator.toString(), `%${value}%`]
    }
}
```

and then, later on

```
const { data } = useHttpQuery('my query', { filter: toHttpFilter(filterModel) })

...
const toHttpFilter = (model: GridFilterModel) => {
   return model.items.map(item => {
     const column = apiRef.current.getColumn(item.field)
     const transformer = column.metadata?.filterTransform
     if (transformer)
        return transformer(item.operator, item.value)
     else
        return [item.field, item.operator, item.value]
   }
}
```

But in short, its just being able to provide additional data about the column thats unique to my application or table.

**Comment by michelengelen** [comment]
Hey @cherniavskii 
This does seem like a nice addition to the type. I would prefer the typescript override variant to align this with the other packages.
WDYT?

**Comment by TeoAvignon** [comment]
Would you please have an update on this ticket?
I think the intention of OP is nice but I do not understand the way to go with this ticket, could you please help?

**Comment by atillaaliyev** [comment]
Do you have any idea about it ? I need same :)

**Comment by ithrforu** [comment]
**Our problem**:
We have a column with gigabytes, but we send bytes to the backend. The maximum number the backend accepts is `Number.MAX_SAFE_INTEGER`. To validate the maximum value in the input of the numeric filter, we must divide `Number.MAX_SAFE_INTEGER` to gigabytes, because when sending to the backend, the reverse process occurs. The unit of measurement can change dynamically. I would like to pass custom field `getMaxGridInputFilterValue` to `GridColDef`, but i cant expand `GridColDef` type. `metadata` field would be useful for us.

CC: @cherniavskii

**Comment by brandonrodgers-wf** [comment]
I agree that this would make life so much easier.  I have a DataGrid built with columns from different sources and it would be nice to be able to add some metadata to the GridColDef with the source name rather than having to look it up everything we need access to it.
